BLD: Set up GDAL_VERSION compile time variable + Add variable to check GDAL GEOS version

jorisvandenbossche opened the pull request.

jorisvandenbossche commented on setup.py:
```diff
@@ -135,6 +135,11 @@ def get_gdal_paths():
 
+    compile_time_env = {
+        "CTE_GDAL_MAJOR_VERSION": 3,
+        "CTE_GDAL_MINOR_VERSION": 4,
```
This of course should be updated to not be hardcoded

brendan-ward commented:
Aha!  This is a clever solution to conditionally bringing in version-specific GDAL features without having to splice together version shim files like Fiona.  On CI, we can easily set a `GDAL_VERSION` (maybe better as a single string that gets parsed) that then gets shadowed with a `CTE_GDAL_VERSION` set internally that is possibly set from `gdal-config` (Linux / MacOS only) if `GDAL_VERSION` is not available.

jorisvandenbossche commented:
I added some code to the `setup.py` (mostly copied from what we had before (which in turn was partly copied from Fiona, I suppose), but we removed that some time ago in https://github.com/geopandas/pyogrio/pull/47).

Some questions:

- It first checks `gdal-config`, and only then falls back to the GDAL_VERSION env variable (this is the way it was done before). In the `get_gdal_paths` functions we actually do this the other way around: give precedence to the env variable. Do we want to do that here as well?
- I copied the fallback to `gdalinfo` on Windows, but we could also just always require an env variable there, to simplify the code
- Fiona also allows to set the version through a command line argument to setup.py: https://github.com/Toblerity/Fiona/blob/9c8fe736d4381526905c6976c399a33d10e3ecbf/setup.py#L132-L138. For Windows we still mention you can do it with CLI arguments as well next to env variables: https://pyogrio.readthedocs.io/en/latest/install.html#windows (not doing that for the version would break that)

brendan-ward commented:
> we actually do this the other way around: give precedence to the env variable. Do we want to do that here as well

I think we should probably be consistent about these; if you are setting the env vars for GDAL when we can automatically discover them anyway via `gdal-config`, we should assume that you intended to override the default.


> just always require an env variable there, to simplify the code

For Windows, builds are involved enough there are not likely to be too many users building from source; let's just require it there.


> Fiona also allows to set the version through a command line argument to setup.py

My instinct is to keep this simple for now, and just use env vars with fallback to `gdal-config` (MacOS / Linux), but since we already support some CLI options it seems like we ought to be consistent with those...

jorisvandenbossche commented:
> I think we should probably be consistent about these; if you are setting the env vars for GDAL when we can automatically discover them anyway via `gdal-config`, we should assume that you intended to override the default.

That sounds good. 
We should then probably (for Linux/Mac) still require that you either set all three of them, or either get all three of them from `gdal-config`? If we do that, I can probably integrate the `get_gdal_version` into `get_gdal_paths` instead of keeping it separate (which has the logic of first checking env variable and then fallback to `gdal-config` twice independently)

brendan-ward commented:
> require that you either set all three of them

Yes, let's make this all-or-nothing.

jorisvandenbossche commented:
OK, updated. (note: the windows part is not tested .. I could do that later, but not really a priority)

brendan-ward commented on setup.py:
```diff
@@ -40,32 +40,35 @@ def read_response(cmd):
 
-    GDAL_INCLUDE_PATH and GDAL_LIBRARY_PATH environment variables are used
-    if both are present.
+    GDAL_INCLUDE_PATH, GDAL_LIBRARY_PATH and GDAL_VERSION environment variables
```
```suggestion
    GDAL_INCLUDE_PATH, GDAL_LIBRARY_PATH, and GDAL_VERSION environment variables
```

brendan-ward commented on setup.py:
```diff
@@ -40,32 +40,35 @@ def read_response(cmd):
         log.warn(
-            "If specifying the GDAL_INCLUDE_PATH or GDAL_LIBRARY_PATH environment "
-            "variables, you need to specify both."
+            "If specifying the GDAL_INCLUDE_PATH, GDAL_LIBRARY_PATH or GDAL_VERSION "
```
```suggestion
            "If specifying the GDAL_INCLUDE_PATH, GDAL_LIBRARY_PATH, or GDAL_VERSION "
```

brendan-ward reviewed (commented):
Thanks for doing this @jorisvandenbossche !

Could you please update `docs/installation.md` to align with this?

I'm not sure how to get Conda / micromamba to give us the GDAL version, but presumably that could be replaced with a matrix variable?

jorisvandenbossche commented:
> I'm not sure how to get Conda / micromamba to give us the GDAL version, but presumably that could be replaced with a matrix variable?

Hmm, that is failing now for Windows. That's actually a reason to maybe keep the gdalinfo fallback on Windows, then we don't need to care about that in this case ... 
But, yes, otherwise we can probably put the version in a matrix variable and update the conda environment to install that. Or actually run `gdalinfo --version` in the CI and set that as env variable together we the others. That might actually be the easiest.

brendan-ward commented:
> run gdalinfo --version in the CI and set that as env variable together we the others

I like this option

jorisvandenbossche commented:
This seems to be working now! (the wheel tests are of course still failing for linux/mac)

jorisvandenbossche commented on .github/workflows/release.yml:
```diff
@@ -226,6 +226,7 @@ jobs:
         run: |
           cd ..
+          python -c "import pyogrio; print(f'GDAL version: {pyogrio.__gdal_version__}\nGEOS version: {pyogrio.__gdal_geos_version__}')"
           python -m pytest --pyargs pyogrio.tests -v
```
For some reason I don't understand, the pytest invocation here doesn't use the `pytest_report_header` from `conftest.py` which reports GDAL version and supported drivers etc (otherwise the explicit print here would not be needed)

brendan-ward replied:
I wonder if adding the `-s` flag would help with that?  Though I thought it should show the header regardless.

jorisvandenbossche commented:
So this unfortunately confirms that GEOS is indeed not enabled for Linux/Mac (but is enabled for Windows) ..

brendan-ward commented on pyogrio/_ogr.pyx:
```diff
@@ -41,6 +41,23 @@ def get_gdal_version_string():
+            return None
+        return (major, minor, revision)
+    ELSE:
+        return (0, 0, 0)
```
I wonder if we should default this to `None` as well?  `None` for GDAL > 3.4 means GEOS is not installed whereas `None` here means we just can't tell.  But really the only thing we can use for debugging is when GEOS is available and GDAL > 3.4.

brendan-ward replied:
I think we either know it or we don't know it.  Since we already can gather GDAL version, returning `(0, 0, 0)` simply shadows that, so I don't think it adds additional clarity.  We'll just have to remember to always be aware of GDAL version.

brendan-ward reviewed (approved):
Thanks @jorisvandenbossche !

One minor suggestion re: GEOS version when GDAL < 3.4.  Feel free to merge at any point.
